include/h2/utils/IntegerMath.hpp

```diff
@@ -0,0 +1,239 @@
+
+template <typename IType>
+inline constexpr auto NBits = IntegerTraits<IType>::nbits;
+
+template <typename IType>
+inline constexpr bool IsSigned = meta::EqV<SType<IType>, IType>();
+
+template <typename IType>
+inline constexpr bool IsUnsigned = meta::EqV<UType<IType>, IType>();
+
+/** @brief Compute `ceil(log_2(d))`.
+ *
+ *  Mathematically this doesn't exist for `d=0`. We define this
+ *  function to return 0 when `d=0`.
+ */
+template <typename IType, typename = meta::EnableWhen<IsUnsigned<IType>>>
+H2_GPU_FORCE_INLINE H2_GPU_HOST_DEVICE auto ceillog2(IType const& d)
+{
+    static constexpr auto nbits = NBits<IType>;
+    int ell = 0;
+    for (ell = 0; ell < nbits; ++ell)
+        if ((static_cast<IType>(1) << ell) >= d)
+            break;
+    return ell;
+}
+
+/** @brief Determine if n is a power of 2. */
+template <typename IType, typename = meta::EnableWhen<IsUnsigned<IType>>>
+H2_GPU_FORCE_INLINE H2_GPU_HOST_DEVICE auto ispow2(IType const& d)
+{
+    return (d & (d - 1)) == 0;
+}
+
+/** @brief Computes the upper 32 bits of `x*y`. */
+H2_GPU_FORCE_INLINE H2_GPU_HOST_DEVICE uint32_t mulhi(uint32_t x,
+                                                      uint32_t y) noexcept
+{
+#if H2_GPU_DEVICE_COMPILING
+    return __umulhi(x, y);
+#else
+    return static_cast<uint32_t>(
+        (static_cast<uint64_t>(x) * static_cast<uint64_t>(y)) >> 32);
+#endif // H2_GPU_DEVICE_COMPILING
+}
+
+/** @brief Computes the upper 64 bits of `x*y`. */
+H2_GPU_FORCE_INLINE H2_GPU_HOST_DEVICE uint64_t mulhi(uint64_t x,
+                                                      uint64_t y) noexcept
+{
+#if H2_GPU_DEVICE_COMPILING
+    return __umul64hi(x, y);
+#else
+    return static_cast<uint64_t>(
+        (static_cast<__uint128_t>(x) * static_cast<__uint128_t>(y)) >> 64);
+#endif // H2_GPU_DEVICE_COMPILING
+}
+
+/** @class FastDiv
+ *  @brief Compute division and mods quickly.
+ *
+ *  The division and mod operations are expensive on pretty much every
+ *  processor, but especially on GPUs. There are ways to work around
+ *  this. This is a persistent container for the "l" and "m'" values
+ *  in Figure 4.1 of [this
+ *  paper](https://gmplib.org/~tege/divcnst-pldi94.pdf).
+ *
+ *  The typical use-case here is breaking down a 1-D index into an N-D
+ *  mutli-index. This class can be used to compute the quotients and
+ *  remainders needed in that calculation. In this case, the constant
+ *  divisor that gets stored is the extent of some dimension. Objects
+ *  of this class implicitly convert back to their IType so that
+ *  extent value does not need to be multiply stored.
+ *
+ *  The setup is all done on the host for performance reasons.
+ *
+ *  @todo Is it worth branching when d=2^l? Shame C++ doesn't have
+ *        dependent typing...
+ */
+template <typename IType, typename = meta::EnableWhen<IsUnsigned<IType>>>
+class FastDiv
+{
+    static_assert(IsUnsigned<IType>, "FastDiv for unsigned division only");
```

Comment:
Is this redundant with the `EnableWhen`?

Reply:
Yes, but it's a compile-time redundancy... I'll remove it.